[CUMULUS-2532] Delete granules before collections and providers

botanical opened the pull request:
**Summary:** Summary of changes

Addresses [CUMULUS-2532: Delete granules before collections and providers ](https://bugs.earthdata.nasa.gov/browse/CUMULUS-2532)

## Changes

* Updated various integration tests to delete granules before providers and collections
* Update some tests to unpublish granules before deletion

## PR Checklist

- [x] Update CHANGELOG
- [ ] Unit tests
- [x] Adhoc testing
- [x] Integration tests

Jkovarik commented on example/spec/parallel/createReconciliationReport/CreateReconciliationReportSpec.js:
```diff
@@ -38,7 +38,7 @@ const { getGranuleWithStatus } = require('@cumulus/integration-tests/Granules');
 const { createProvider } = require('@cumulus/integration-tests/Providers');
 const { deleteCollection, getCollections } = require('@cumulus/api-client/collections');
-const { deleteGranule } = require('@cumulus/api-client/granules');
+const { deleteGranule, removePublishedGranule } = require('@cumulus/api-client/granules');
```
Nit: When bringing in the api-client we should consider sorting moving all of the  granulesApiTestUtils calls to use `api-client`.   Scope is a fine reason not to in this PR, however.

Jkovarik replied:
Nit: When bringing in the api-client we should consider sorting moving all of the granulesApiTestUtils calls to use api-client. Scope is a fine reason not to in this PR, however.

Jkovarik commented on example/spec/parallel/dbRecords/DbRecordsDLQSpec.js:
```diff
@@ -29,7 +29,7 @@ describe('When a bad record is sent on the DLQ', () => {
     const granuleId = randomString(10);
     const files = [fakeFileFactory()];
-    const granule = fakeGranuleFactoryV2({ files, granuleId });
+    const granule = fakeGranuleFactoryV2({ files, granuleId, published: false });
```
This change generally makes sense, but what specifically required it?

botanical replied:
Attempting to delete a published granule results in an error https://github.com/nasa/cumulus/blob/master/packages/api/endpoints/granules.js#L207

Jkovarik replied:
Ah, right, it's an endpoint constraint based on the key, not an actual CMR value.    👍🏻  Thanks for the clarification.

Jkovarik commented on example/spec/parallel/discoverGranules/DiscoverGranulesDuplicateHandlingSkipSpec.js:
```diff
@@ -197,10 +197,16 @@ describe('The DiscoverGranules workflow with one existing granule, one new granu
         () => deleteGranule({ prefix, granuleId: existingGranuleId }),
         () => deleteGranule({ prefix, granuleId: newGranuleId }),
+      ],
+      { stopOnError: false }
```
@npauzenga noting as this is one of the :thumbsup: patterns I was talking about instead of the Promise.all setups a lot of the tests are using.

Jkovarik commented on example/spec/parallel/discoverGranules/DiscoverGranulesHttpsSpec.js:
```diff
@@ -79,22 +87,21 @@ describe('The Discover Granules workflow with https Protocol', () => {
         'DiscoverGranules'
       );
-      lambdaOutput = await lambdaStep.getStepOutput(
+      discoverGranulesLambdaOutput = await lambdaStep.getStepOutput(
```
:thumbsup:  Nice change!

Jkovarik commented on example/spec/parallel/discoverGranules/DiscoverGranulesHttpsSpec.js:
```diff
@@ -137,23 +144,34 @@ describe('The Discover Granules workflow with https Protocol', () => {
     });
 
     describe('SyncGranule lambda function', () => {
       it('outputs 1 granule', async () => {
```
super-minor-nit: consider updating to something like 'outputs the expected granule'

botanical replied:
[`815243d` (#2254)](https://github.com/nasa/cumulus/pull/2254/commits/815243d4a145724b531e1380b1140e63f738bb96)

Jkovarik commented on example/spec/parallel/discoverGranules/DiscoverGranulesWithExecutionNamePrefixSpec.js:
```diff
@@ -37,14 +42,16 @@ describe('The DiscoverGranules workflow', () => {
     process.env.ProvidersTable = `${stackName}-ProvidersTable`;
 
-    const testId = randomString();
+    // const testId = randomString();
```
Remove comment please!

botanical replied:
Ah, true! [`4a16bd0` (#2254)](https://github.com/nasa/cumulus/pull/2254/commits/4a16bd0e8a7d1c976b08aedec207154590d72d09)

Jkovarik commented on example/spec/parallel/discoverGranules/DiscoverGranulesWithExecutionNamePrefixSpec.js:
```diff
@@ -94,8 +101,10 @@ describe('The DiscoverGranules workflow', () => {
+  afterAll(async () => {
+    await waitForCompletedExecution(workflowExecution.executionArn);
+    await granulesApiTestUtils.deleteGranule({ prefix: stackName, granuleId: 'MOD09GQ.A2016358.h13v04.006.2016360104606' });
+    await Promise.all([
```
I commented on #2252 as well, I think we need to consider a Promise.allSettled or pMap with stop on failure set to false for a bunch of these, so as much is cleaned up as possible, even if there's a failure.     Obviously out of scope for this PR, but raising for visibility/future discussion.

botanical replied:
[`815243d` (#2254)](https://github.com/nasa/cumulus/pull/2254/commits/815243d4a145724b531e1380b1140e63f738bb96)

Jkovarik replied:
Right, pAll is what I meant.   💯

Jkovarik commented on example/spec/parallel/ingest/ingestFromPdrSpec.js:
```diff
@@ -256,9 +256,15 @@ describe('Ingesting from PDR', () => {
         await granulesApiTestUtils.deleteGranule({
           prefix: config.stackName,
-          granuleId: parseLambdaOutput.payload.granules[0].granuleId,
+          granuleId: testDataGranuleId,
```
Shouldn't this be covered by the statement on L259?

botanical replied:
Yeah, thanks for finding that!

botanical replied:
[`815243d` (#2254)](https://github.com/nasa/cumulus/pull/2254/commits/815243d4a145724b531e1380b1140e63f738bb96)

Jkovarik commented on example/spec/parallel/ingest/ingestFromPdrWithExecutionNamePrefixSpec.js:
```diff
@@ -63,16 +65,17 @@ describe('The DiscoverAndQueuePdrsExecutionPrefix workflow', () => {
   const providersDir = './data/providers/s3/';
   const collectionsDir = './data/collections/s3_MOD09GQ_006';
 
+  let addedCollection;
```
:100: !

Jkovarik commented on example/spec/parallel/ingest/ingestFromPdrWithExecutionNamePrefixSpec.js:
```diff
@@ -167,7 +176,7 @@ describe('The DiscoverAndQueuePdrsExecutionPrefix workflow', () => {
     if (beforeAllFailed) fail('beforeAll() failed');
     else {
-      const executionArn = queuePdrsOutput.payload.running[0];
+      executionArn = queuePdrsOutput.payload.running[0];
```
Should we be setting the 'global' test variable here and on L190?     `executionArn` looks like a value that should be set in the beforeAll where queuePdrsOutput is obtained.

botanical replied:
Yep! [`89cd3a5` (#2254)](https://github.com/nasa/cumulus/pull/2254/commits/89cd3a567602e004c381fa52223a15ec3540b02b)

Jkovarik commented on example/spec/parallel/ingestGranule/IngestGranuleFailureSpec.js:
```diff
@@ -84,8 +85,13 @@ describe('The Ingest Granule failure workflow', () => {
 
-      // delete the granule record from DynamoDB if exists
-      await granuleModel.delete({ granuleId: inputPayload.granules[0].granuleId });
+      await deleteGranule({ prefix: config.stackName, granuleId: inputPayload.granules[0].granuleId });
```
I'll circle back to this, but it's not immediately clear why granules[0] is explicitly called before the map on the granules object.

botanical replied:
Hm that is a great point. I will remove the map since it's not needed but I am not sure why we need to delete the granule record if we are making it fail by adding a non existent file to it.

botanical replied:
[`815243d` (#2254)](https://github.com/nasa/cumulus/pull/2254/commits/815243d4a145724b531e1380b1140e63f738bb96)

Jkovarik replied:
👀    Hmm, you're right.   This test is validating a failure, so pre-cleanup isn't *really* required, particularly with the other changes to granule cleanup we're making. 

Interestingly enough, we're *indexing* in the cleanup which isn't great if we changed that fixture, but I'm not incredibly concerned re: that for now.

Jkovarik commented on example/spec/parallel/ingestGranule/IngestGranuleWithDuplicateHandlingVersionSpec.js:
```diff
@@ -238,12 +238,12 @@ describe('The IngestGranule workflow with DuplicateHandling="version" and a gran
       { stopOnError: false }
     ).catch(console.error);
 
+    await deleteGranule({ prefix, granuleId });
```
Nit/non-blocking: Should be be try/caching this sort of deletion 'hoisting' and doing a console.error like what results from the other calls?      It's probably irrelevant as jasmine will print out AfterAll failures, but I hadn't really deeply thought about this pattern before now.

Jkovarik commented on packages/api/src/lib/granule-delete.ts:
```diff
@@ -62,7 +62,7 @@ const deleteGranuleAndFiles = async ({
   granuleModelClient: typeof Granule
 }) => {
-  logger.info(`Deleting PG granule ${pgGranule} mapped to dynamoGranule ${dynamoGranule}`);
+  logger.info(`Deleting PG granule ${JSON.stringify(pgGranule)} mapped to dynamoGranule ${JSON.stringify(dynamoGranule)}`);
```
👍🏻

Jkovarik reviewed (changes requested):
This all looks really good!  Nice work!  Just a couple of really minor cleanup/questions and this should be good to go!

botanical commented:
@Jkovarik Thanks for the review! I think I addressed the feedback you've left me apart from the non-blocking comment which I hope to get to if time permits.

Jkovarik commented on example/spec/parallel/discoverGranules/DiscoverGranulesHttpsSpec.js:
```diff
@@ -67,6 +68,12 @@ describe('The Discover Granules workflow with https Protocol', () => {
   afterAll(async () => {
     // clean up stack state added by test
+    await Promise.all(discoverGranulesLambdaOutput.payload.granules.map(
+      (granule) => granulesApiTestUtils.deleteGranule({
```
Nit: This method is deprecated.  We shoudn't be using it, instead bring in the api-client granule delete method directly.

Jkovarik commented on example/spec/parallel/discoverGranules/DiscoverGranulesHttpsSpec.js:
```diff
@@ -79,21 +86,20 @@ describe('The Discover Granules workflow with https Protocol', () => {
     afterAll(async () => {
-      await Promise.all(lambdaOutput.payload.granules.map(
+      await Promise.all(discoverGranulesLambdaOutput.payload.granules.map(
         (granule) => granulesApiTestUtils.deleteGranule({
```
Nit: same as above

Jkovarik commented on example/spec/parallel/ingestGranule/IngestUMMGSuccessSpec.js:
```diff
@@ -213,6 +213,12 @@ describe('The S3 Ingest Granules workflow configured to ingest UMM-G', () => {
   afterAll(async () => {
     // clean up stack state added by test
+    await Promise.all(inputPayload.granules.map(
+      (g) => granulesApiTestUtils.removePublishedGranule({
```
Nit: this is deprecated, we should be using the api-client methods directly

botanical replied:
Definitely agree, thank you for this call out. I will handle this in this PR and comment back with the commit 👍

Jkovarik replied:
Sure thing -- to be clear I'm fine with punting this to scoped work/a ticket to wholesale remove these calls (they're all over in our testing).

Jkovarik reviewed (approved):
Approved, good work.   👍🏻 

If you have time, consider changing the calls to `granuleApiTestUtils` should be using the api-client method directly.

Jkovarik commented on example/spec/parallel/discoverGranules/DiscoverGranulesWithExecutionNamePrefixSpec.js:
```diff
@@ -97,22 +107,32 @@ describe('The DiscoverGranules workflow', () => {
 
   afterAll(async () => {
+    await Promise.all(
+      queueGranulesOutput.payload.running
```
I noted while re-reviewing -- If the purpose of waiting for completed execution is simply to delete the final granule write, we should probably be using the executions API to wait for completion instead of the step functions API (which this helper ultimately uses), *then* check for granule status, *then* delete, if we want to be sure we're using the database state to determine status.

Jkovarik commented on example/spec/parallel/testAPI/sqsRuleSpec.js:
```diff
@@ -92,6 +93,9 @@ async function cleanUp() {
 
   // TODO there may be another execution to delete here
```
Is this still a valid TODO?

botanical replied:
I believe so. I've made a note about it in the next ticket

Jkovarik reviewed (approved):
On re-review, this all looks great, I had one additional comment we may want to consider capturing, but in every way this update is an improvement to the test cleanup/logging.   Good work!
